add guide for writing and deploying LEZ programs with logos-scaffold- #406 (#410)
status-im/ift-docs: docs/lez/programs/write-and-deploy-lez-program-with-scaffold.md

---
title: Write and deploy an LEZ program with `logos-scaffold`
doc_type: procedure
product: blockchain
topics:
  - lez
  - scaffold
steps_layout: sectioned
authors: ygd58, kashepavadan
owner: logos
doc_version: 1
slug: write-and-deploy-lez-program-with-scaffold
sidebar_position: 1
---

# Write and deploy an LEZ program with `logos-scaffold`

#### Use `logos-scaffold` to create, build, and deploy a guest program on the Logos Execution Zone testnet.

[`logos-scaffold`](https://github.com/logos-co/scaffold) is a project scaffold and CLI tool that manages the full lifecycle of a [LEZ](../../get-started/glossary.md#lez) guest [program](../../get-started/glossary.md#program) — from project creation to deployment. It pins LEZ and SPEL dependencies, builds a project-local sequencer, and handles wallet interactions, so you can focus on writing your program logic.

Before you begin, ensure you have:

- Linux x86\_64 or macOS
- Rust and Cargo (latest stable)
- Node.js v20 or later

> - A running [Logos Blockchain](../../get-started/glossary.md#logos-blockchain) node connected to the public testnet (see [Run a Logos Blockchain node from Basecamp](../../blockchain/get-started/run-a-logos-blockchain-node-from-basecamp.md))

Reviewer: **Blocker + inaccurate prerequisites** — found while dogfooding this guide end-to-end on Linux x86_64. Four problems in this list:

1. **Missing: the RISC Zero toolchain.** `logos-scaffold build` (Step 5) fails immediately without it:
   ```
   Risc Zero Rust toolchain not found. Try running `rzup install rust`
   ```
   And once it builds, the sequencer *crashes* while executing a deployed program (`Sequencer failed unexpectedly: Main loop exited unexpectedly`) unless `r0vm` is also installed. Bootstrap:
   ```bash
   curl -L https://risczero.com/install | bash
   rzup install rust   # required by `logos-scaffold build`
   rzup install r0vm   # required by the sequencer to execute transactions
   ```
2. **Missing: a container runtime.** `logos-scaffold doctor` requires Docker or Podman for the guest build (the generated project's own README lists *"Docker or Podman (required by guest method builds)"*).
3. **`Node.js v20 or later` is unused** — there is no Node/npm anywhere in `logos-scaffold` or the generated project. Safe to drop.
4. **`A running Logos Blockchain node …` is incorrect.** The scaffold is standalone — its README states *"No logos-blockchain dependency"* / *"Standalone sequencer flow only"*, and Step 3 builds a project-local sequencer. Requiring a separate blockchain node here misleads readers.

Suggested replacement for the list:

```suggestion
- Linux x86\_64 or macOS
- `git`, `curl`, and Rust/Cargo (latest stable)
- Docker or Podman — required for the RISC0 guest method build
- The [RISC Zero toolchain](https://dev.risczero.com/api/zkvm/install) installed via `rzup`: run `rzup install rust` (needed by `logos-scaffold build`) and `rzup install r0vm` (needed by the sequencer to execute transactions)
```

---
_Generated by [Claude Code](https://claude.ai/code)_

Reviewer: this comment seems correct as it follows https://github.com/logos-blockchain/logos-execution-zone#install-build-dependencies

2. Inspect the project layout:

    ```text
    my-program/
    ├── scaffold.toml          # Project configuration and dependency pins
    ├── methods/
    │   └── guest/
    │       └── src/bin/       # Guest programs run inside the RISC0 zkVM
    ├── src/
    │   └── bin/               # Runner scripts that submit transactions
    └── .scaffold/             # Local state, wallet home, and build artifacts
    ```

## Step 3: Set up the project

1. Run `setup` to sync the LEZ and SPEL repositories to their pinned commits, build the project-local sequencer and wallet binaries, and seed the default wallet:

    ```bash
    logos-scaffold setup
    ```

    This step can take several minutes on a cold cache as it builds the sequencer from source.

## Step 4: Write your guest program

Guest programs run inside the [RISC0 zkVM](https://dev.risczero.com/) and define the on-chain logic of your LEZ program. Each guest program in `methods/guest/src/bin/` becomes a deployable program with its own `program_id`.

1. Open the sample guest program:

    ```bash
    $EDITOR methods/guest/src/bin/hello_world.rs
    ```

> 1. The program receives an `Input` struct via the zkVM environment, applies your logic, and writes an `Output` struct to the journal. The sequencer verifies the proof and updates the on-chain [account](../../get-started/glossary.md#account) state.

Reviewer: Minor accuracy: the sample `hello_world.rs` reads/writes `ProgramInput` / `ProgramOutput` (from `nssa_core::program`), not bare `Input` / `Output` types. Matching the code readers will open:

```suggestion
1. The program receives a `ProgramInput` struct via the zkVM environment, applies your logic, and writes a `ProgramOutput` struct to the journal. The sequencer verifies the proof and updates the on-chain [account](../../get-started/glossary.md#account) state.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

    Key concepts:
    - **Instructions** are encoded as `Vec<u8>` (opcode byte followed by payload).
    - **Account data** is stored in `AccountWithMetadata` structs.
    - Use `RISC0_DEV_MODE=1` during development to skip ZK proof generation for faster iteration.

## Step 5: Build the project

1. Build the workspace. In development, use `RISC0_DEV_MODE=1` to skip proof generation:

    ```bash
    RISC0_DEV_MODE=1 logos-scaffold build
    ```

>     The build compiles your guest programs and produces `.bin` artifacts under `.scaffold/`.

Reviewer: The build does not place `.bin` artifacts under `.scaffold/`. In a clean run they land under the guest target dir — e.g. `target/riscv-guest/<methods-crate>/<programs>/riscv32im-risc0-zkvm-elf/release/*.bin`. `.scaffold/` holds the circuits, wallet, state, and logs. (Confirmed: `deploy` submits the ELF from exactly that `target/riscv-guest/...` path.)

```suggestion
    The build compiles your guest programs and produces `.bin` artifacts under `target/riscv-guest/…/riscv32im-risc0-zkvm-elf/release/`.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

## Step 6: Start a local sequencer

1. Start a project-local sequencer to test your program before deploying to the testnet:

    ```bash
    RISC0_DEV_MODE=1 logos-scaffold localnet start
    ```

    The sequencer is daemonised and survives terminal or tmux session closure. Use `logos-scaffold localnet status` to check that it is running and `logos-scaffold localnet stop` to stop it.

## Step 7: Deploy your program

1. Deploy all guest programs to the running sequencer:

    ```bash
    RISC0_DEV_MODE=1 logos-scaffold deploy
    ```

>     After a successful deployment, `logos-scaffold` prints the `program_id` for each guest program — a hex-encoded RISC0 image ID computed from the submitted ELF. Save the `program_id`; you need it to interact with your program.

Reviewer: In a clean dogfood, `deploy` printed `program_id: unavailable (run logos-scaffold setup to build the vendored spel)` for **every** program — even though `setup` had completed successfully and `doctor` reported the `spel` binary present. So the promise that a `program_id` is always printed and that you *need* it to interact doesn't hold on the happy path: the Step 8 runner scripts load the program from its embedded ELF, not from a `program_id` you paste in. Suggest softening, and confirming with the scaffold maintainers whether the `unavailable` output is a bug:

```suggestion
    After a successful deployment, `logos-scaffold` prints a per-program summary; when the vendored `spel` tooling is available it also prints a `program_id` — a hex-encoded RISC0 image ID computed from the submitted ELF. The example runner scripts in Step 8 load the program from its embedded ELF, so you do not need to copy a `program_id` to complete this guide.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

1. To deploy a specific program by name:

    ```bash
    RISC0_DEV_MODE=1 logos-scaffold deploy hello_world
    ```

## Step 8: Interact with your program

Use the project-local wallet CLI to submit transactions to your deployed program. The wallet is available at `logos-scaffold wallet`.

> 1. Check your wallet balance:

Reviewer: Two issues: (a) this needs the **sequencer from Step 6 running** — `wallet topup` does an account preflight against `http://127.0.0.1:3040` and errors (`sequencer appears unavailable …`) if it isn't up; (b) `wallet list` lists **accounts**, not balances, so "Check your wallet balance" mislabels the step (`topup` is a faucet claim).

```suggestion
1. With the sequencer from Step 6 running, top up the default wallet from the faucet, then list your accounts (`wallet list` shows accounts, not a balance):
```

---
_Generated by [Claude Code](https://claude.ai/code)_

    ```bash
    logos-scaffold wallet topup
    logos-scaffold wallet list
    ```

1. Run one of the example runner scripts that submit transactions to your program:

    ```bash

>     RISC0_DEV_MODE=1 cargo run --bin run_hello_world

Reviewer: **Blocker.** As written this fails two ways in a clean run:

- `error: the following required arguments were not provided: <ACCOUNT_ID>` — the runner takes a self-owned public account id (from `logos-scaffold wallet list`).
- Even with the argument, a direct `cargo run` panics: `Failed to read persistent storage at /root/.nssa/wallet/storage.json … Not found` — the runner locates the wallet via `NSSA_WALLET_HOME_DIR`, which the `logos-scaffold wallet` passthrough sets automatically but a direct `cargo run` does not. The generated project's own skill docs and README confirm both points (`cargo run --bin run_hello_world -- $PUBLIC_HELLO_ACCOUNT_ID`).

With both fixes I got `submitted transaction: tx_hash=…` and the sequencer included it in a block. Suggested fix:

```suggestion
    export NSSA_WALLET_HOME_DIR="$(pwd)/.scaffold/wallet"
    RISC0_DEV_MODE=1 cargo run --bin run_hello_world -- <PUBLIC_ACCOUNT_ID>
```
where `<PUBLIC_ACCOUNT_ID>` is a self-owned `Public/…` account from `wallet list`.

---
_Generated by [Claude Code](https://claude.ai/code)_

---
title: Write and deploy an LEZ program with `logos-scaffold`
doc_type: procedure
product: blockchain
topics:
  - lez
  - scaffold
steps_layout: sectioned
authors: ygd58, kashepavadan
owner: logos
doc_version: 1
slug: write-and-deploy-lez-program-with-scaffold
sidebar_position: 1
---

# Write and deploy an LEZ program with `logos-scaffold`

#### Use `logos-scaffold` to create, build, and deploy a guest program on the Logos Execution Zone testnet.

[`logos-scaffold`](https://github.com/logos-co/scaffold) is a project scaffold and CLI tool that manages the full lifecycle of a [LEZ](../../get-started/glossary.md#lez) guest [program](../../get-started/glossary.md#program) — from project creation to deployment. It pins LEZ and SPEL dependencies, builds a project-local sequencer, and handles wallet interactions, so you can focus on writing your program logic.

Before you begin, ensure you have:

- Linux x86\_64 or macOS
- `git`, `curl`, and Rust/Cargo (latest stable)
- Docker or Podman — required for the RISC0 guest method build

> - The [RISC Zero toolchain](https://dev.risczero.com/api/zkvm/install) installed via `rzup`: run `rzup install rust` (needed by `logos-scaffold build`) and `rzup install r0vm` (needed by the sequencer to execute transactions)

Reviewer: The prerequisites are missing **Nix**. `scaffold.toml` pins `basecamp` and `lgpm` with `build = "nix-flake"`, and `logos-scaffold doctor` on a fresh project reports:

```
WARN | tool nix | not found on PATH
  remediation: Install `nix`
```

So `logos-scaffold setup` can't build those dependencies without Nix. Add it to the list:

```suggestion
- The [RISC Zero toolchain](https://dev.risczero.com/api/zkvm/install) installed via `rzup`: run `rzup install rust` (needed by `logos-scaffold build`) and `rzup install r0vm` (needed by the sequencer to execute transactions)
- [Nix](https://nixos.org/download/) — `logos-scaffold setup` builds the `basecamp` and `lgpm` dependencies as Nix flakes (`build = "nix-flake"` in `scaffold.toml`); `logos-scaffold doctor` flags a missing `nix` with *Install `nix`*
```
